File: src/beatsync/wasm/audio_to_clock.cpp
#include "audio_to_clock.h"
#include <cmath>
#include <iostream>
#include <numeric>
#include <chrono>

// No Emscripten-specific includes here

AudioToClock::AudioToClock(const AudioToClockConfig& config,
                           OnPredictionReadyCallback onPredictionReady,
                           OnRequestFeatureExtractionCallback onRequestFeatureExtraction,
                           OnRequestBpmPhasePredictionCallback onRequestBpmPhasePrediction,
                           OnBestTrajectoryCallback onBestTrajectory,
                           OnStabilizerDebugDataCallback onStabilizerDebugData,
                           OnExternalClockAdjustedCallback onExternalClockAdjusted,
                           OnExternalClockDebugDataCallback onExternalClockDebugData)
    : inferenceManager_(
          config.inferenceConfig,
          [this, onPredictionReady](const BpmPhasePredictorResult& result, const InferenceManagerDebugData& debugData) {
            stabilizer_.addPrediction(result.phase, result.phaseMagnitude, result.bpm, result.inputTime);
            onPredictionReady(result, debugData);
          },
          onRequestFeatureExtraction,
          onRequestBpmPhasePrediction
      ),
      stabilizer_(
          config.stabilizerConfig,
          [this, onBestTrajectory](const StabilizerTrajectory& bestTrajectory) {
              externalClockController_.update(
                bestTrajectory.bpm,
                Stabilizer::predictPhase(bestTrajectory, latestTimestamp_),
                Stabilizer::predictBarPhase(bestTrajectory, latestTimestamp_),
                latestTimestamp_);
              onBestTrajectory(bestTrajectory);
          },
          onStabilizerDebugData
      ),
      externalClockController_(
          config.externalClockControllerConfig,
          onExternalClockAdjusted,
          onExternalClockDebugData
      )
{
    // No onStatusUpdated here, as it's WASM-specific
}

void AudioToClock::SetForceExportAllDebugData(bool forceExport) {
    inferenceManager_.SetForceExportAllDebugData(forceExport);
    stabilizer_.SetForceExportAllDebugData(forceExport);
    externalClockController_.SetForceExportAllDebugData(forceExport);
}

void AudioToClock::addAudio(uintptr_t audioSamplesPtr, int numChannels, int numSamples, double currentTime, int inputSampleRate) {
    latestTimestamp_ = currentTime;
    float* audioSamples = reinterpret_cast<float*>(audioSamplesPtr);

    std::vector<std::vector<float>> channels(numChannels);
    for (int i = 0; i < numChannels; ++i) {
        channels[i].resize(numSamples);
        std::copy(audioSamples + i * numSamples, audioSamples + (i + 1) * numSamples, channels[i].begin());
    }

    inferenceManager_.addAudio(channels, currentTime, inputSampleRate);
    inferenceManager_.process();
}

void AudioToClock::resolveFeatureExtractor(uintptr_t odfDataPtr, size_t odfLength, uintptr_t specDataPtr, size_t specLength) {
    FeatureExtractorResult res;
    float* odfData = reinterpret_cast<float*>(odfDataPtr);
    float* specData = reinterpret_cast<float*>(specDataPtr);
    res.odf.assign(odfData, odfData + odfLength);
    res.spec.assign(specData, specData + specLength);
    inferenceManager_.onFeaturesReady(res);
}

void AudioToClock::resolveBpmPhasePredictor(const BpmPhasePredictorResult& result) {
    inferenceManager_.onPredictionReady(result);
}

void AudioToClock::tick(double currentTime) {
    externalClockController_.tick(currentTime);
}

void AudioToClock::resync() {
    externalClockController_.resync();
    stabilizer_.resync();
}

double AudioToClock::getBarPhase() const {
    return externalClockController_.getBarPhase();
}

File: src/beatsync/audio_to_clock.ts
import { ExternalClockController } from "./external_clock_controller";
import { AudioToClockConfig, IAudioToClock, IAudioToClockConstructor } from "./schema";
import { InferenceManager, InferenceManagerConfigInternal } from "./inference_manager";
import { Stabilizer } from "./stabilizer";
import { SAMPLE_RATE, BLOCK_DURATION_S, HI_RES_HOP_LENGTH, LOW_RES_HOP_LENGTH, LOW_RES_N_FFT } from './config_audio';
import { predictPhase, predictBarPhase } from "./extrapolation";

class AudioToClock implements IAudioToClock {
  private inferenceManager: InferenceManager | null = null;
  private readonly stabilizer: Stabilizer;
  private readonly externalClockController: ExternalClockController;
  private latestTime = 0.0;

  constructor(config: AudioToClockConfig) {
    const inferenceConfig: InferenceManagerConfigInternal = {
      ...config.inferenceConfig,
      onPrediction: ({ bpm, phase, phaseMagnitude, inputTime, debugData }) => {
        this.stabilizer.addPrediction(phase, phaseMagnitude, bpm, inputTime);
        config.onDebugDataExported?.({ inference: debugData });
      },
      runFeatureExtractor: config.runFeatureExtractor,
      runBpmPhasePredictor: config.runBpmPhasePredictor,
    } as InferenceManagerConfigInternal;
    this.inferenceManager = new InferenceManager(inferenceConfig);

    this.stabilizer = new Stabilizer({
      ...config.stabilizerConfig,
      onTrajectoryUpdated: (bestTrajectory) => {
        const currentTime = this.latestTime;
        this.externalClockController.update(
          bestTrajectory.bpm,
          predictPhase(bestTrajectory, currentTime),
          predictBarPhase(bestTrajectory, currentTime),
          currentTime,
        );
      },
      onDebugDataUpdated: (debugData) => {
        config.onDebugDataExported?.({ stabilizer: debugData });
      }
    });

    this.externalClockController = new ExternalClockController({
      ...config.externalClockControllerConfig,
    });
    this.externalClockController.onClockAdjusted((changes) => {
      config.onExternalClockAdjusted?.({ bpm: changes.bpm, phase: changes.phase, timestamp: changes.timestamp });
      config.onDebugDataExported?.({ externalClock: changes.debugData, externalClockEvent: changes });
    });
    this.externalClockController.start();
  }

  addAudio(audioSamples: Float32Array[], currentTime: number, inputSampleRate: number) {
    this.latestTime = currentTime;
    this.inferenceManager?.addAudio(audioSamples, currentTime, inputSampleRate);
  }

  resync(hard?: boolean): void {
    // No-op for now
  }

  resetHardSync(): void {
    // No-op for now
  }

  setForceExportAllDebugData(force: boolean): void {
    // No-op for now
  }
}

export const AudioToClockConstructor: IAudioToClockConstructor = AudioToClock;
